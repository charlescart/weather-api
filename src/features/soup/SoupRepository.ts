/* eslint-disable no-param-reassign */
import soupDto from './validators/soupDto';
import summaryDto from './validators/summaryDto';

export default class SoupRepository {
  static Soup({ soup, searchWords }: soupDto): summaryDto {
    const summary: summaryDto = this.CreateSummary(searchWords);

    soup.forEach((row: string[], rowIndex: number) => {
      const numberColumns: number = row.length - 1;
      row.forEach((letter: string, columnIndex: number) => {
        // verificar de solo a una letra
        this.LookUpOneLetterWords(letter.toUpperCase(), summary);

        // mirando hacia la derecha
        this.LookToTheRight(columnIndex, numberColumns, summary, row);

        // mirando hacia la diagonal inferior derecha
        this.LookAtDiagonal(rowIndex, columnIndex, summary, soup);

        // mirando hacia abajo
        this.LookAtDown(soup, rowIndex, columnIndex, summary);

        // diagonal inferior izquierda
        this.LookAtDiagonalLeft(rowIndex, columnIndex, summary, soup);

        // mirando hacia la derecha
        this.LookAtLeft(soup, rowIndex, columnIndex, summary);

        // mirando hacia la diagonal superior izquierda
        this.LookAtTopLeftDiagonal(soup, rowIndex, columnIndex, summary);

        // mirando hacia arriba
        this.LookAtTop(soup, rowIndex, columnIndex, summary);

        // mirando hacia la diagonal superior derecha
        this.LookTopRightDiagonal(soup, rowIndex, columnIndex, summary);
      });
    });

    return summary;
  }

  static LookUpOneLetterWords(letter: string, summary: summaryDto): summaryDto {
    const wordsSizeOne: string[] = Object.keys(summary)
      .filter((word) => word.length === 1);

    for (let i = 0; i < wordsSizeOne.length; i += 1) {
      const word: string = wordsSizeOne[i];
      if (letter === word) summary[word] += 1;
    }

    return summary;
  }

  static LookToTheRight(columnIndex: number, columns: number, summary: summaryDto, row: string[]): summaryDto {
    const spaceToTheRight: number = columns - columnIndex;
    const words: string[] = Object.keys(summary)
      .filter((word) => (word.length - 1) >= 1 && (word.length - 1) <= spaceToTheRight);

    for (let i = 0; i < words.length; i += 1) {
      const word: string = words[i];
      const limit: number = (columnIndex + spaceToTheRight);
      const phrase: string = row
        .filter((item, index) => index >= columnIndex && index <= limit)
        .join('')
        .toUpperCase();

      if (word === phrase.substr(0, word.length)) summary[word] += 1;
    }

    return summary;
  }

  static LookAtDiagonal(rowIndex: number, columnIndex: number, summary: summaryDto, soup: string[][]): summaryDto {
    const numberRows: number = soup.length;
    let numberColumns: number = soup[rowIndex].length;
    let phraseDiagonal = '';

    for (let i = rowIndex; i < numberColumns; i += 1) {
      for (let j = columnIndex; j < numberColumns && i < numberRows; j += 1) {
        phraseDiagonal += soup[i][j].toUpperCase();
        i += 1;
        if (soup[i] !== undefined) numberColumns = soup[i].length;
      }
      break;
    }

    phraseDiagonal = phraseDiagonal.trim();

    if (phraseDiagonal.length > 1) {
      this.SearchPhraseMatch(phraseDiagonal, summary);
    }

    return summary;
  }

  static LookAtDown(soup: string[][], row: number, column: number, summary: summaryDto): summaryDto {
    let phrase = '';

    for (let i = row; i < soup.length; i += 1) {
      phrase += soup[i][column].toUpperCase();
      if (soup[i + 1] === undefined || (soup[i + 1].length - 1) < column) break;
    }

    phrase = phrase.trim();

    if (phrase.length > 1) {
      this.SearchPhraseMatch(phrase, summary);
    }

    return summary;
  }

  static LookAtDiagonalLeft(row: number, column: number, summary: summaryDto, soup: string[][]): summaryDto {
    let phrase = '';

    if (column > 0) {
      let i = row;
      for (let j = column; j >= 0 && i < soup.length; j -= 1) {
        phrase += soup[i][j].toUpperCase();
        i += 1;
        if (soup[i] !== undefined && soup[i].length < j) break;
      }

      phrase = phrase.trim();
      if (phrase.length > 1) this.SearchPhraseMatch(phrase, summary);
    }

    return summary;
  }

  static LookAtLeft(soup: string[][], row: number, column: number, summary: summaryDto): summaryDto {
    let phrase = '';

    for (let i = column; i >= 0; i -= 1) {
      phrase += soup[row][i].toUpperCase();
    }

    phrase = phrase.trim();
    if (phrase.length > 1) this.SearchPhraseMatch(phrase, summary);

    return summary;
  }

  static LookAtTopLeftDiagonal(soup: string[][], row: number, column: number, summary: summaryDto): summaryDto {
    let phrase = '';

    if (row && column > 0) {
      let i = row;
      for (let j = column; i >= 0 && j >= 0; j -= 1) {
        phrase += soup[i][j].toUpperCase();
        i -= 1;
      }
      phrase = phrase.trim();
    }

    if (phrase.length > 1) this.SearchPhraseMatch(phrase, summary);
    return summary;
  }

  static LookAtTop(soup: string[][], row: number, column: number, summary: summaryDto): summaryDto {
    let phrase = '';

    for (let i = row; i >= 0 && row > 0; i -= 1) {
      if ((soup[i].length - 1) < column) break;
      phrase += soup[i][column].toUpperCase();
    }

    phrase = phrase.trim();
    if (phrase.length > 1) this.SearchPhraseMatch(phrase, summary);
    return summary;
  }

  static LookTopRightDiagonal(soup: string[][], row: number, column: number, summary: summaryDto): summaryDto {
    let phrase = '';

    let j = column;
    for (let i = row; i >= 0 && row > 0; i -= 1) {
      phrase += soup[i][j].toUpperCase();
      j += 1;
      if (soup[i - 1] !== undefined && (soup[i - 1].length - 1) < j) break;
    }

    phrase = phrase.trim();
    if (phrase.length > 1) this.SearchPhraseMatch(phrase, summary);
    return summary;
  }

  static CreateSummary(words: string[]): summaryDto {
    const summary: summaryDto = {};

    for (let i = 0; i < words.length; i += 1) {
      const word: string = words[i];
      if (word.length > 0) summary[word.toUpperCase()] = 0;
    }

    return summary;
  }

  static SearchPhraseMatch(phrase: string, summary: summaryDto): summaryDto {
    const words: string[] = Object.keys(summary)
      .filter((word) => word.length > 1 && word.length <= phrase.length);

    words.forEach((word) => {
      if (word === phrase.substr(0, word.length)) summary[word] += 1;
    });

    return summary;
  }
}
