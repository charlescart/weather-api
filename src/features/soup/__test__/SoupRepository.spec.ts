import SoupRepository from '../SoupRepository';
import Summary from '../validators/summaryDto';

describe('SoupRepository', () => {
  it('CreateSummary', () => {
    const words: string[] = ["oiE", "ixE", "Ix", "i", "O", "awds", "e4", "", "oi", "eX"];
    let summary: Summary = SoupRepository.CreateSummary(words);
    expect(summary).toMatchSnapshot();
  });

  it('LookUpOneLetterWords', () => {
    const words: string[] = ['o', 'I', 'e', 'X', '4', 'f', 'k', 'p', '1', '2', '3'];
    let summary: Summary = SoupRepository.CreateSummary(words);
    const soup: string[][] = [
      ["o", "I", "E"],
      ["I", "i", "X", "4"],
      ["E", "x", "E"],
      ["E", "X"],
      ["e"],
      [""]
    ];

    soup.forEach((row: string[]) => {
      row.forEach((letter: string) => {
        letter = letter.toUpperCase();
        SoupRepository.LookUpOneLetterWords(letter, summary);
      });
    });

    expect(summary).toMatchSnapshot();
  });

  it('LookToTheRight', () => {
    const words: string[] = ["oiE", "Oi", "ie", "ii", "eXe", "EX", "XE", "X4", "IIX4", "", "palabraSuperGrande"];
    let summary: Summary = SoupRepository.CreateSummary(words);
    const soup: string[][] = [
      ["o", "I", "E"],
      ["I", "i", "X", "4"],
      ["E", "x", "E"],
      ["E", "X"],
      ["e"],
      [""]
    ];

    soup.forEach((row: string[], rowIndex: number) => {
      const numberColumns: number = row.length - 1;
      row.forEach((letter: string, columnIndex: number) => {
        letter = letter.toUpperCase();
        SoupRepository.LookToTheRight(columnIndex, numberColumns, summary, row);
      });
    });

    expect(summary).toMatchSnapshot();
  });

  it('LookAtDiagonal', () => {
    const words: string[] = ["palabraSuperGrande", "oie", "OI", "iX", "E4", "Ex", ""];
    let summary: Summary = SoupRepository.CreateSummary(words);
    const soup: string[][] = [
      ["o", "I", "E"],
      ["I", "i", "X", "4"],
      ["E", "x", "E"],
      ["E", "X"],
      ["e"],
      [""]
    ];

    soup.forEach((row: string[], rowIndex: number) => {
      const numberColumns: number = row.length - 1;
      row.forEach((letter: string, columnIndex: number) => {
        letter = letter.toUpperCase();
        SoupRepository.LookAtDiagonal(rowIndex, columnIndex, summary, soup);
      });
    });

    expect(summary).toMatchSnapshot();
  });

  it('LookAtDown', () => {
    const words: string[] = ["palabraSuperGrande", "oie", "OI", "oieee", "IIXX", "ii", "IiX", "iX", "E4", "Ex", "E5", "4", "EE", "xx"];
    let summary: Summary = SoupRepository.CreateSummary(words);
    const soup: string[][] = [
      ["o", "I", "E"],
      ["I", "i", "X", "4"],
      ["E", "X", "E"],
      ["E", "X", "E", "5"],
      ["E"],
      [" "]
    ];

    soup.forEach((row: string[], rowIndex: number) => {
      row.forEach((letter: string, columnIndex: number) => {
        SoupRepository.LookAtDown(soup, rowIndex, columnIndex, summary);
      });
    });

    expect(summary).toMatchSnapshot();
  });

  it('SearchPhraseMatch', () => {
    const phrases: string[] = ["OIEEE", "IIXX", "EXEE", "IEEE", "IXX", "XEE", "EEE", "XX", "EE", "EE"];
    const words: string[] = ["palabraSuperGrande", "oie", "OI", "oieee", "IIXX", "ii", "IiX", "iX", "E4", "Ex", "E5", "4", "EE", "xx"];
    const summary: Summary = SoupRepository.CreateSummary(words);

    phrases.forEach(phrase => SoupRepository.SearchPhraseMatch(phrase, summary));

    expect(summary).toMatchSnapshot();
  });

  it('LookAtDiagonalLeft', () => {
    const words: string[] = ["palabraSuperGrande", "ii", "EIE", "EI", "Ei", "eie", "ie", "xx", "XXE", "4E", "4EX", "4exe", "xe", "exe"];
    let summary: Summary = SoupRepository.CreateSummary(words);
    const soup: string[][] = [
      ["o", "I", "E"],
      ["I", "i", "X", "4"],
      ["E", "X", "E"],
      ["E", "X", "E", "5"],
      ["E"],
      [" "]
    ];

    soup.forEach((row: string[], rowIndex: number) => {
      row.forEach((letter: string, columnIndex: number) => {
        SoupRepository.LookAtDiagonalLeft(rowIndex, columnIndex, summary, soup);
      });
    });

    expect(summary).toMatchSnapshot();
  });

  it('LookAtLeft', () => {
    const words: string[] = ["palabraSuperGrande", "eio", "ei", "IO", "4XII", "xII", "ii", "exe", "xE", "5exe", "1", "34", "awd"];
    let summary: Summary = SoupRepository.CreateSummary(words);
    const soup: string[][] = [
      ["o", "I", "E"],
      ["I", "i", "X", "4"],
      ["E", "X", "E"],
      ["E", "X", "E", "5"],
      ["E"],
      [" "]
    ];

    soup.forEach((row: string[], rowIndex: number) => {
      row.forEach((letter: string, columnIndex: number) => {
        SoupRepository.LookAtLeft(soup, rowIndex, columnIndex, summary);
      });
    });

    expect(summary).toMatchSnapshot();
  });

  it('LookAtTopLeftDiagonal', () => {
    const words: string[] = ["aws", "io", "xi", "4e", "EIO", "EI", "5EIO", "XE", "EXI", "ex", "5E", "", "2", "23424"];
    let summary: Summary = SoupRepository.CreateSummary(words);
    const soup: string[][] = [
      ["o", "I", "E"],
      ["I", "i", "X", "4"],
      ["E", "X", "E"],
      ["E", "X", "E", "5"],
      ["E"],
      [" "]
    ];

    soup.forEach((row: string[], rowIndex: number) => {
      row.forEach((letter: string, columnIndex: number) => {
        SoupRepository.LookAtTopLeftDiagonal(soup, rowIndex, columnIndex, summary);
      });
    });

    expect(summary).toMatchSnapshot();
  });

  it('LookAtTop', () => {
    const words: string[] = ["awd", "io", "ii", "XE", "4", "EI", "EIO", "xi", "xII", "EXE", "EX", "EEIO", "EEEio", "XXII", "EEXE"];
    let summary: Summary = SoupRepository.CreateSummary(words);
    const soup: string[][] = [
      ["o", "I", "E"],
      ["I", "i", "X", "4"],
      ["E", "X", "E"],
      ["E", "X", "E", "5"],
      ["E"],
      [" "]
    ];

    soup.forEach((row: string[], rowIndex: number) => {
      row.forEach((letter: string, columnIndex: number) => {
        SoupRepository.LookAtTop(soup, rowIndex, columnIndex, summary);
      });
    });

    expect(summary).toMatchSnapshot();
  });

  it('LookTopRightDiagonal', () => {
    const words: string[] = ["awd", "II", "IE", "x", "4", "EIE", "XX", "E4", "EXX", "XE4", "EXE4", "e", "5", "", "exe"];
    let summary: Summary = SoupRepository.CreateSummary(words);
    const soup: string[][] = [
      ["o", "I", "E"],
      ["I", "i", "X", "4"],
      ["E", "X", "E"],
      ["E", "X", "E", "5"],
      ["E"],
      [" "]
    ];

    soup.forEach((row: string[], rowIndex: number) => {
      row.forEach((letter: string, columnIndex: number) => {
        SoupRepository.LookTopRightDiagonal(soup, rowIndex, columnIndex, summary);
      });
    });

    expect(summary).toMatchSnapshot();
  });
});