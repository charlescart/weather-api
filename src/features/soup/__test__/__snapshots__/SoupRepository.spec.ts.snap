// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SoupRepository CreateSummary 1`] = `
Object {
  "AWDS": 0,
  "E4": 0,
  "EX": 0,
  "I": 0,
  "IX": 0,
  "IXE": 0,
  "O": 0,
  "OI": 0,
  "OIE": 0,
}
`;

exports[`SoupRepository LookAtDiagonal 1`] = `
Object {
  "E4": 1,
  "EX": 1,
  "IX": 2,
  "OI": 1,
  "OIE": 1,
  "PALABRASUPERGRANDE": 0,
}
`;

exports[`SoupRepository LookAtDiagonalLeft 1`] = `
Object {
  "4E": 1,
  "4EX": 1,
  "4EXE": 1,
  "EI": 1,
  "EIE": 1,
  "EXE": 1,
  "IE": 1,
  "II": 1,
  "PALABRASUPERGRANDE": 0,
  "XE": 2,
  "XX": 1,
  "XXE": 1,
}
`;

exports[`SoupRepository LookAtDown 1`] = `
Object {
  "4": 0,
  "E4": 0,
  "E5": 0,
  "EE": 3,
  "EX": 1,
  "II": 1,
  "IIX": 1,
  "IIXX": 1,
  "IX": 1,
  "OI": 1,
  "OIE": 1,
  "OIEEE": 1,
  "PALABRASUPERGRANDE": 0,
  "XX": 1,
}
`;

exports[`SoupRepository LookAtLeft 1`] = `
Object {
  "1": 0,
  "34": 0,
  "4XII": 1,
  "5EXE": 1,
  "AWD": 0,
  "EI": 1,
  "EIO": 1,
  "EXE": 2,
  "II": 1,
  "IO": 1,
  "PALABRASUPERGRANDE": 0,
  "XE": 2,
  "XII": 1,
}
`;

exports[`SoupRepository LookAtTop 1`] = `
Object {
  "4": 0,
  "AWD": 0,
  "EEEIO": 2,
  "EEIO": 1,
  "EEXE": 1,
  "EI": 1,
  "EIO": 1,
  "EX": 1,
  "EXE": 1,
  "II": 1,
  "IO": 1,
  "XE": 1,
  "XI": 1,
  "XII": 1,
  "XXII": 1,
}
`;

exports[`SoupRepository LookAtTopLeftDiagonal 1`] = `
Object {
  "2": 0,
  "23424": 0,
  "4E": 1,
  "5E": 1,
  "5EIO": 1,
  "AWS": 0,
  "EI": 1,
  "EIO": 1,
  "EX": 1,
  "EXI": 1,
  "IO": 1,
  "XE": 1,
  "XI": 2,
}
`;

exports[`SoupRepository LookToTheRight 1`] = `
Object {
  "EX": 2,
  "EXE": 1,
  "IE": 1,
  "II": 1,
  "IIX4": 1,
  "OI": 1,
  "OIE": 1,
  "PALABRASUPERGRANDE": 0,
  "X4": 1,
  "XE": 1,
}
`;

exports[`SoupRepository LookTopRightDiagonal 1`] = `
Object {
  "4": 0,
  "5": 0,
  "AWD": 0,
  "E": 0,
  "E4": 1,
  "EIE": 1,
  "EXE": 1,
  "EXE4": 1,
  "EXX": 1,
  "IE": 1,
  "II": 1,
  "X": 0,
  "XE4": 1,
  "XX": 1,
}
`;

exports[`SoupRepository LookUpOneLetterWords 1`] = `
Object {
  "1": 0,
  "2": 0,
  "3": 0,
  "4": 1,
  "E": 5,
  "F": 0,
  "I": 3,
  "K": 0,
  "O": 1,
  "P": 0,
  "X": 3,
}
`;

exports[`SoupRepository SearchPhraseMatch 1`] = `
Object {
  "4": 0,
  "E4": 0,
  "E5": 0,
  "EE": 3,
  "EX": 1,
  "II": 1,
  "IIX": 1,
  "IIXX": 1,
  "IX": 1,
  "OI": 1,
  "OIE": 1,
  "OIEEE": 1,
  "PALABRASUPERGRANDE": 0,
  "XX": 1,
}
`;
